File: pp-exercises/mpi/exercise3/ring-datatype.f90
module module_my_struct

  type my_struct
     real(8) :: x
     integer :: i
  end type my_struct
end


program main  ! ring with derived datatype

  use module_my_struct
  implicit none
  include 'mpif.h'
  integer :: size, rank, ierr
  integer :: partner_left, partner_right
  integer :: len(2), type(2), my_type
  integer(MPI_ADDRESS_KIND) :: disp(2)

  type(my_struct) :: sum, sbuf, rbuf

  call mpi_init(ierr)
  call mpi_comm_size(MPI_COMM_WORLD, size, ierr)
  call mpi_comm_rank(MPI_COMM_WORLD, rank, ierr)

  partner_left  = mod(rank - 1 + size, size)
  partner_right = mod(rank + 1, size)

  len(1) = 1
  len(2) = 1

  call mpi_get_address(sum%x, disp(1), ierr)
  call mpi_get_address(sum%i, disp(2), ierr)
 
  disp(2) = disp(2) - disp(1)
  disp(1) = disp(1) - disp(1)

  type(1) = MPI_REAL8
  type(2) = MPI_INTEGER

  call mpi_type_create_struct(2, len, disp, type, my_type, ierr)
  call mpi_type_commit(my_type, ierr)
  
  sum%x = 0
  sum%i = 0
  sbuf%x = rank
  sbuf%i = rank

  do 
     call mpi_sendrecv(sbuf, 1, my_type, partner_right, 0, &
                       rbuf, 1, my_type, partner_left, 0, &
                       MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)
     sum%x = sum%x + rbuf%x
     sum%i = sum%i + rbuf%i
     sbuf%x = rbuf%x
     sbuf%i = rbuf%i
     if (rbuf%i == rank) exit
  enddo  

  write(6,*) "sums on rank ", rank, ": ", sum%x, sum%i

  call mpi_finalize(ierr)
end

File: pp-exercises/mpi/exercise3/ring-datatype.c
#include <stdio.h>
#include <mpi.h>

typedef struct myStruct { double x; int i; } myStruct;

int main(int argc, char *argv[])  // ring with derived datatype
{
    int size, rank;
    int partner_left, partner_right;
    int len[2];
    myStruct sum, sbuf, rbuf;
    MPI_Aint disp[2];
    MPI_Datatype type[2];
    MPI_Datatype myType;

    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    partner_left  = (rank - 1 + size) % size;
    partner_right = (rank + 1) % size;

    len[0] = 1;
    len[1] = 1;

    MPI_Get_address(&sum.x, &disp[0]);
    MPI_Get_address(&sum.i, &disp[1]);
 
    disp[1] -= disp[0];
    disp[0] -= disp[0];

    type[0] = MPI_DOUBLE;
    type[1] = MPI_INT;

    MPI_Type_create_struct(2, len, disp, type, &myType);
    MPI_Type_commit(&myType);

    sum.x = 0;
    sum.i = 0;
    sbuf.x = rank;
    sbuf.i = rank;

    do {
	MPI_Sendrecv(&sbuf, 1, myType, partner_right, 0,
		     &rbuf, 1, myType, partner_left, 0, 
                     MPI_COMM_WORLD, MPI_STATUS_IGNORE);
	sum.x += rbuf.x;
	sum.i += rbuf.i;
	sbuf.x = rbuf.x;
	sbuf.i = rbuf.i;
    } while (rbuf.i != rank);
  
    printf("sums on rank %d: %f, %d\n", rank, sum.x, sum.i);
    
    MPI_Finalize();
    return 0;
}  
